handlers.go

```diff
@@ -181,6 +181,9 @@ func (dht *IpfsDHT) handlePutValue(ctx context.Context, p peer.ID, pmes *pb.Mess
 		return nil, err
 	}
 
+	dht.putLock.Lock()
+	defer dht.putLock.Unlock()
```

Comment:
This lock needs to move _above_ the get to make the put atomic. That is, we need to ensure that the new value is strictly better than the old one. With the current logic, we can:

1. Read the value.
2. Someone else can update the value.
3. Take the lock.
4. Compare against the _old_ value.
5. See that we have a newer one.
6. Put the value, overwriting the potentially _even newer_ value form step 2.

Reply:
@Stebalien  Have pushed the changes for the striped locking. Let me know what you think.